README.rst
==========================
``gs.group.messages.text``
==========================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Support for displaying the plain-text version of a post
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Author: `Michael JasonSmith`_
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2015-10-30
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.net`_.

..  _Creative Commons Attribution-Share Alike 4.0 International License:
    http://creativecommons.org/licenses/by-sa/4.0/

Introduction
============

This product provides the utilities and functions that supports
the displaying of these posts, particularly the conversion of the
plain-text post to HTML.

The **actual** **rendering** of the messages is carried out in
either:

#.  `gs.group.messages.post.text.base`_ for the plain-text
    version of the post shown on the web, or
#.  `gs.group.list.email.html`_ for the HTML version of the
    plain-text message (the *pseudo* *HTML*) that is used in
    email messages.

Resources
=========

- Documentation
  https://groupserver.readthedocs.org/projects/gsgroupmessagestext
- Code repository:
  https://github.com/groupserver/gs.group.messages.text/
- Questions and comments to
  http://groupserver.org/groups/development
- Report bugs at https://redmine.iopen.net/projects/groupserver

.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net
.. _Michael JasonSmith: http://groupserver.org/p/mpj17

.. _gs.group.messages.post.text.base:
   https://github.com/groupserver/gs.group.messages.post.text.base
.. _gs.group.list.email.html:
   https://github.com/groupserver/gs.group.list.email.html


docs/HISTORY.rst
Changelog
=========

1.1.0 (2016-03-31)
------------------

* Adding the quotation symbols (``> ``) to wrapped lines.
* Tweaking the bottom-quoting detection code so messages from
  Mozilla Thunderbird (like `this message`_ are handled
  correctly.

.. _this message:
   http://groupserver.org/r/post/4l8yWaMqinDBwtyqiZ3ZvU

1.0.0 (2015-10-30)
------------------

Initial version. Prior to the creation of this product the code
for handling the plain-text version of the posts was split
between `gs.group.messages.post.text`_ and
`gs.group.list.email.html`_.

.. _gs.group.messages.post.text:
   https://github.com/groupserver/gs.group.messages.post.text

.. _gs.group.list.email.html:
   https://github.com/groupserver/gs.group.list.email.html

..  LocalWords:  Changelog iframe


gs/group/messages/text/tests/matcher.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2015 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from unittest import TestCase
from gs.group.messages.text.matcher import (
    Matcher, boldMatcher, emailMatcher, wwwMatcher, uriMatcher)


class TestMatcher(TestCase):
    def test_match(self):
        re = '.*(fish).*'
        m = Matcher(re, None)

        r = m.match('I am a fish.')
        self.assertTrue(r)

    def test_miss(self):
        'Test when they do not match'
        re = '.*(fish).*'
        m = Matcher(re, None)

        r = m.match('I am a dog.')
        self.assertFalse(r)

    def test_sub(self):
        re = 'fish'
        sub = 'dog'
        m = Matcher(re, sub)

        r = m.sub('I am a fish.')
        self.assertEqual('I am a dog.', r)


class TestBoldMatcher(TestCase):
    def test_match(self):
        'Test that "*this*" is recognised as bold'
        r = boldMatcher.match("*this*")
        self.assertTrue(r)

    def test_sub(self):
        r = boldMatcher.sub('I am a *fish.*')
        self.assertEqual('I am a <b>*fish.*</b>', r)


class TestEmailMatcher(TestCase):
    def test_match(self):
        'Test that "person@example.com" is recognised as an email address'
        r = emailMatcher.match("person@example.com")
        self.assertTrue(r)

    def test_sub(self):
        'Test email substitution'
        r = emailMatcher.sub('person@example.com')
        expected = '<a class="email" href="mailto:person@example.com">person@example.com</a>'
        self.assertEqual(expected, r)

    def test_An_Email_match(self):
        'Test email markup with mixed case'
        r = emailMatcher.match('A.Person@Example.com')
        self.assertTrue(r)

    def test_An_Email_sub(self):
        'Test email substitution with mixed case'
        r = emailMatcher.sub('A.Person@Example.com')
        self.assertEqual(
            '<a class="email" href="mailto:A.Person@Example.com">A.Person@Example.com</a>', r)

    def test_angle_email_match(self):
        'Test when we match when Michael writes an email address'
        r = emailMatcher.match('<person@example.com>')
        self.assertTrue(r)

    def test_angle_email_sub(self):
        'Test when we substitute when Michael writes an email address'
        r = emailMatcher.sub('&lt;person@example.com&gt;')
        expected = '<a class="email" href="mailto:person@example.com">&lt;person@example.com&gt;'\
            '</a>'
        self.assertEqual(expected, r)

    def test_mailto_email_match(self):
        'Test email address written as a mailto-URI is matched'
        r = emailMatcher.match('mailto:person@example.com')
        self.assertTrue(r)

    def test_mailto_email_sub(self):
        'Test email address written as a mailto-URI is substituted'
        r = emailMatcher.sub('mailto:person@example.com')
        self.assertEqual(
            '<a class="email" href="mailto:person@example.com">mailto:person@example.com</a>', r)


class TestWWWMatcher(TestCase):
    def test_www_match(self):
        'Test that www is matched'
        r = wwwMatcher.match('www.example.com')
        self.assertTrue(r)

    def test_www_sub(self):
        'Test that www is turned into a link'
        r = wwwMatcher.sub('www.example.com')
        self.assertEqual('<a href="http://www.example.com">www.example.com</a>', r)

    def test_WWW_match(self):
        'Test that WWW is matched'
        r = wwwMatcher.match('WWW.Example.Com')
        self.assertTrue(r)

    def test_WWW_sub(self):
        'Test that WWW is turned into a link'
        r = wwwMatcher.sub('WWW.Example.Com')
        self.assertEqual('<a href="http://WWW.Example.Com">WWW.Example.Com</a>', r)


class TestURIMatcher(TestCase):
    def test_http_match(self):
        'Test that an http-address is matched as a URI'
        r = uriMatcher.match('http://example.com')
        self.assertTrue(r)

    def test_http_sub(self):
        'Test that an http-address is turned into a link'
        r = uriMatcher.sub('http://example.com')
        self.assertEqual('<a href="http://example.com">http://<b>example.com</b></a>', r)

    def test_https_match(self):
        'Test that an https-address is matched as a URI'
        r = uriMatcher.match('https://example.com')
        self.assertTrue(r)

    def test_https_sub(self):
        'Test that an https-address is turned into a link'
        r = uriMatcher.sub('https://example.com')
        self.assertEqual('<a href="https://example.com">https://<b>example.com</b></a>', r)

    def test_http_path_match(self):
        'Test a http-address with a path is matched'
        r = uriMatcher.match('http://example.com/people/me')
        self.assertTrue(r)

    def test_add_zws(self):
        'Test the adding zero-width spaces to a URL, to allow for breaking of long URLs'
        t = '/a/long/path'
        r = uriMatcher.add_zws(t)
        self.assertEqual('&#8203;/a&#8203;/long&#8203;/path', r)
        self.assertEqual('/a/long/path', t)  # Ensure nothing changes with the orignal string

    def test_http_path_sub(self):
        'Test a http-address with a path is turned into a link'
        r = uriMatcher.sub('http://example.com/people/me')
        expected = '<a href="http://example.com/people/me">http://<b>example.com</b>/people/me</a>'
        self.assertEqual(expected, r)

    def test_http_query_match(self):
        'Test an http-address with a query string is matched'
        r = uriMatcher.match('http://example.com/people/me?show=Stufff')
        self.assertTrue(r)

    def test_http_query_sub(self):
        'Test an http-address with a query string is turned into a link'
        r = uriMatcher.sub('http://example.com/people/me?show=Stufff')
        expected = '<a href="http://example.com/people/me?show=Stufff">'\
            'http://<b>example.com</b>&#8203;/people&#8203;/me&#8203;?show&#8203;=Stufff</a>'
        self.assertEqual(expected, r)

    def test_http_angle_match(self):
        'Test an http-address in angle brackets matches'
        r = uriMatcher.match('&lt;http://example.com/&gt;')
        self.assertTrue(r)

    def test_http_angle_sub(self):
        'Test an http-address in angle brackets is turned into a link'
        r = uriMatcher.sub('&lt;http://example.com/&gt;')
        expected = '<a href="http://example.com/">&lt;http://<b>example.com</b>/&gt;</a>'
        self.assertEqual(expected, r)

    def test_long_https_angle_match(self):
        'Test a long https-address in angle brackets is matched'
        text = '&lt;https://groups.example.com/people/a_very_long_user_id?show=Stufff&gt;'
        r = uriMatcher.match(text)
        self.assertTrue(r)

    def test_long_https_angle_sub(self):
        'Test a long https-address in angle brackets is turned into a link'
        text = '&lt;https://groups.example.com/people/a_very_long_user_id?show=Stufff&gt;'
        r = uriMatcher.sub(text)
        expected = '<a class="small" href="https://groups.example.com/people/'\
            'a_very_long_user_id?show=Stufff">&lt;https://<b>groups.example.com</b>&#8203;/people'\
            '&#8203;/a&#8203;_very&#8203;_long&#8203;_user&#8203;_id&#8203;?show&#8203;=Stufff&gt;'\
            '</a>'
        self.assertEqual(expected, r)


gs/group/messages/text/tests/__init__.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, unicode_literals, print_function
#lint:disable
from .htmlbody import TestHTMLBody
from .matcher import TestMatcher
from .matcher import TestBoldMatcher
from .matcher import TestEmailMatcher
from .matcher import TestWWWMatcher
from .matcher import TestURIMatcher
from .splitmessage import TestSplitMessage
from .test_all import load_tests
#lint:enable


gs/group/messages/text/htmlbody.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2015 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from operator import attrgetter
from xml.sax.saxutils import escape
from zope.cachedescriptors.property import Lazy
from .matcher import (boldMatcher, emailMatcher, wwwMatcher, uriMatcher, )
from .splitmessage import split_message


class HTMLBody(object):
    '''The HTML form of a plain-text email body.

:param str originalText: The original (plain) text'''
    HTML_ESCAPE_TABLE = {
        '"': "&quot;",
        "'": "&apos;"
    }

    def __init__(self, originalText):
        if not originalText:
            raise(ValueError('"originalText" argument required'))
        self.originalText = originalText
        self.matchers = [boldMatcher, emailMatcher, wwwMatcher, uriMatcher]
        sorted(self.matchers, key=attrgetter('weight'))

    def __iter__(self):
        '''The marked-up lines in the main body'''
        mainBody = self.splitBody[0]
        lines = mainBody.rstrip().split('\n')
        for line in lines:
            retval = self.markup(line)
            yield retval

    def __unicode__(self):
        '''The main part of the HTML body, as a Unicode string'''
        retval = '\n'.join(self)
        return retval

    def __str__(self):
        '''The main part of the HTML body, as an ASCII string. Non-ASCII characters are replaced
with XML entities.'''
        retval = unicode(self).encode('ascii', 'xmlcharrefreplace')
        return retval

    @Lazy
    def splitBody(self):
        '''The body as a 2-tuple: main body, and remainder'''
        retval = split_message(self.originalText)
        return retval

    def markup(self, line):
        '''Markup the line, and the words in the line

:param str line: The line to mark up.
:returns: An HTML form of the line: the characters escaped, the words marked up, and surrounded
          in a ``<span>`` element.
:rtype: str'''
        if line.strip() == '':
            retval = '&#160;<br/>'
        else:
            cssClass = "line"
            # The ">From" is a Unix from, so the line is not a quote
            if ((line.lstrip()[0] == '>') and (line.lstrip()[:5] != '>From')):
                cssClass += " muted"
            #  <https://wiki.python.org/moin/EscapingHtml>
            escapedLine = escape(line.rstrip(), self.HTML_ESCAPE_TABLE)
            markedUpLine = self.markup_words(escapedLine)
            r = '<span class="{0}">{1}</span><br/>'
            retval = r.format(cssClass, markedUpLine)
        assert(retval)
        return retval

    def markup_words(self, line):
        '''Mark up the words on the line

:param str line: The line to mark up
:returns: The line with the words marked up
:rtype: str'''
        rwords = []
        for word in line.split(' '):
            subWord = None  # Word that will be substituted for the current word
            # Short-circut if the word is ''. It will be turned back in ' ' when we ``' '.join``
            if word:
                for matcher in self.matchers:
                    if matcher.match(word):
                        subWord = matcher.sub(word)
                        break
            rword = subWord if subWord is not None else word
            rwords.append(rword)
        retval = ' '.join(rwords)
        return retval


gs/group/messages/text/postbody.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2012, 2013, 2014, 2015, 2016 OnlineGroups.net and
# Contributors.
#
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from re import compile as re_compile
from textwrap import TextWrapper

# this is currently the hard limit on the number of word's we will process.
# after this we insert a message. TODO: make this more flexible by using
# AJAX to incrementally fetch large emails
EMAIL_WORD_LIMIT = 5000


class Wrapper(object):
    #: What to break the words on. It is based on the one inside the
    #: :class:`textwrap.TextWrapper` class, but without the breaking on '-'.
    splitExp = re_compile(r'(\s+|(?<=[\w\!\"\'\&\.\,\?])-{2,}(?=\w))')

    #: How to identify quotes
    quoteExp = re_compile(r'(>+\ +)')

    def __init__(self, width):
        self.email_wrapper = TextWrapper(
            width=width, expand_tabs=False, replace_whitespace=False, break_on_hyphens=False,
            break_long_words=False)
        self.email_wrapper.wordsep_re = self.splitExp

        self.quote_wrapper = TextWrapper(
            expand_tabs=False, replace_whitespace=False, break_on_hyphens=False,
            break_long_words=False)
        self.quote_wrapper.wordsep_re = self.splitExp

    def wrap_line(self, l):
        m = self.quoteExp.match(l)
        if m:
            # Quotation wrap
            quoteChars = m.group(1)
            self.quote_wrapper.subsequent_indent = quoteChars
            retval = self.quote_wrapper.fill(l)
        else:
            # Normal wrap
            retval = self.email_wrapper.fill(l)
        return retval


def wrap_message(messageText, width=79):
    """Word-wrap the message

:param str messageText: The text to alter.
:param int width: The column-number which to wrap at.
:returns: The wrapped text.
:rtype: str

.. Note: Originally a stand-alone script in
         ``Presentation/Tofu/MailingListManager/lscripts``."""
    wrapper = Wrapper(width)
    filledLines = [wrapper.wrap_line(l) for l in messageText.split('\n')]
    retval = '\n'.join(filledLines)
    return retval


gs/group/messages/text/splitmessage.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2012, 2013, 2014, 2015 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from collections import (deque, namedtuple)
from re import (compile as re_compile, )

#: The 2-tuple containing the strings representing
#:
#: 0. The main body of the message (``intro``) and
#: 1. The rest of the message, including the bottom-quoting and the footer (``remainder``).
SplitMessage = namedtuple('SplitMessage', ['intro', 'remainder'])

#: The regular expression for the **title** **element** **contents.** For example,
#: ``Post by Dinsdale Piranha: Violence: British gangland: Ethel the Frog``. If we see this then
#: an email client, such as Mozilla Thunderbird, has decided to bottom quote, and convert the
#: contents of the ``<head>`` element to plain text. This would be fine if it was just the
#: ``<title>`` element, but it will be followed by some rather ugly CSS from some ``<style>``
#: elements.
#:
#: The four components seperated by a colon are the author-name, topic-name, group name, and
#: site name. *At least* three groups sperated by colons are expected, but there could be more if
#: any name contains a colon itself.
postByRE = re_compile('^\s*Post by (.*:){3,}')

#: The strings that are commonly used to explicitly indicate a signiture starting
EXPLICIT_SIG_START = ['--', '==', '~~', '__']


def split_message(messageText, max_consecutive_comment=12, max_consecutive_whitespace=3):
    """Split the message into main body and the footer.

:param str messageText: The text to process.
:param int max_consecutive_comment: The maximum number of lines of quoting to allow before snipping.
:param int max_consecutive_whitespace: The maximum number of lines that just contain whitespace to
    allow before snipping.
:returns: 2-tuple, containing the strings representing the main-body of the message, and the footer.
:rtype: :class:`SplitMessage`

Email messages often contain a footer at the bottom, which identifies the user, and who they work
for. However, GroupServer has lovely profiles which do this, so normally we want to snip the footer,
to reduce clutter.

In addition, many users only write a short piece of text at the top of the email, while the
remainder of the message consists of all the previous posts. This method also removes the
*bottom quoting*.

Originally a ZMI-side script in ``Presentation/Tofu/MailingListManager/lscripts``."""
    intro = []
    remainder = deque()
    remainder_start = False
    consecutive_comment = 0
    consecutive_whitespace = 0

    for i, line in enumerate(messageText.split('\n'), 1):
        if ((line[:2] in EXPLICIT_SIG_START) or (line[:3] == '- -') or postByRE.match(line)):
            remainder_start = True

        if remainder_start:
            # If we've started on the remainder, just append to the remainder
            remainder.append(line)
        elif (consecutive_comment > max_consecutive_comment) and (i > 25):
            # Add comments (quotes) to the remainder, but don't penalise top-quoting
            remainder.append(line)
            remainder_start = True
        elif (i <= 15):
            # if we've got less than 15 lines, just put it in the intro
            intro.append(line)
        elif (len(line) > 3) and ((line[:4] != '&gt;') or (line[:2] != '> ')):
            intro.append(line)
        elif consecutive_whitespace < max_consecutive_whitespace:
            # It is < (rather than <=) because of how we count the lines
            intro.append(line)
        else:
            remainder.append(line)
            remainder_start = True

        # Raise a flag if we see a quote (comment). This is used in the if-statement above
        if (line and ((((line[:4] == '&gt;') or (line[:1] == '>')) and
                       ((line[:9] != '&gt;From ') or line[:6] != '>From '))
                      or (line.lower().find('wrote:') != -1))):
            consecutive_comment += 1
        else:
            consecutive_comment = 0
        # Raise a flag if we see a whitespace. This is used in the major if-statement above
        if len(line.strip()):
            consecutive_whitespace = 0
        else:
            consecutive_whitespace += 1

    # Backtrack through the intro, in reverse order, adding things to either the remainder or
    # keeping them in the intro
    rintro = deque()
    trim = len(intro) > 5
    for i, line in enumerate(intro[::-1]):
        if trim:
            ls = line[:4] if len(line) > 3 else ''
            if (((ls == '&gt;') or (ls[:2] == '> ') or (ls.strip() == '')
                 or (line.find('wrote:') > 2))):
                # IF we are trimming, and we are looking at a quote-character or an empty string
                #   OR we have seen the 'wrote:' marker
                # THEN add it to the snipped-text.
                remainder.appendleft(line)
            else:
                trim = False
                rintro.appendleft(line)
        else:
            rintro.appendleft(line)

    # Do not snip, if we will only snip a single line of actual content
    if (len(remainder) == 1) or ((len(remainder) == 2) and ('' in remainder)):
        rintro.extend(remainder)
        remainder = []

    intro = '\n'.join(rintro).strip()
    remainder = '\n'.join(remainder)

    # If we have snipped too much
    if not intro:
        intro = remainder
        remainder = ''

    retval = SplitMessage(intro, remainder)
    return retval


gs/group/messages/text/matcher.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2015 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals, print_function
from re import compile as re_compile, I as re_I, M as re_M, U as re_U
from string import punctuation


class Matcher(object):
    '''Match a word, by a regular expression, and make a substitution

:param str matchRE: The regular expression used to check if there was a match
                    (see :func:`re.match`)
:param str subStr: The string specifying the subsitution (see :func:`re.sub`)'''
    def __init__(self, matchRE, subStr, weight=10):
        self.matchRE = matchRE
        self.subStr = subStr
        self.weight = weight

        #: The regular expression used to make the match. The flags :const:`re.I`, :const:`re.M`,
        #: and :const:`re.U` are set.
        self.re = re_compile(self.matchRE, re_I | re_M | re_U)

    def match(self, s):
        '''Does the string match the regular expression?

:param str s: The string to evaluate
:returns: ``True`` if the string matches the regular expression, ``False`` otherwise.
:rtype: bool'''
        return self.re.match(s)

    def sub(self, s):
        '''Substitute the string in for the substitution string

:param str s: The string to process
:returns: The new string substituted in :attr:`self.subStr`
:rtype: unicode'''
        return self.re.sub(self.subStr, s)

#: Turn words within ``*asterisk*`` characters into bold-elements. This is as close as
#: GroupServer gets to implementing a wiki.
boldMatcher = Matcher("(?P<boldText>\*.*\*)", r'<b>\g<boldText></b>', 10)

#: Turn email addresses (``person@example.com``) into clickable ``mailto:`` links. Surrounding
#: text (such as parenthesis) is added to the *link text*, while the address is extractd and used
#: as for the *link target*.
emailMatcher = Matcher(
    r"(?P<leading>.*?)(?P<address>[A-Z0-9\._%+-]+@[A-Z0-9.-]+\.[A-Z]+)(?P<trailing>.*)",
    r'<a class="email" href="mailto:\g<address>">\g<leading>\g<address>\g<trailing></a>', 20)

#: Turn site names that start with *www* (``www.example.com``) into clickable ``http://`` links.
wwwMatcher = Matcher(r"(?P<siteName>www\..+)",
                     r'<a href="http://\g<siteName>">\g<siteName></a>', 30)


class URIMatcher(Matcher):
    '''A horrid hack for a horrid issue'''
    def __init__(self):
        super(URIMatcher, self).__init__(
            r"(?P<leading>\&lt;|\(|\[|\{|\"|\'|^)"
            r"(?P<protocol>http://|https://)"
            r"(?P<host>([a-z\d][-a-z\d]*[a-z\d]\.)*[a-z][-a-z\\d]+[a-z])"
            r"(?P<rest>.*?)"
            r"(?P<trailing>\&gt;|\)|\]|\}|\"|\'|$|\s)",
            r'<a href="\g<protocol>\g<host>\g<rest>">\g<leading>\g<protocol><b>\g<host></b>'
            r'\g<rest>\g<trailing></a>', 40)

    def sub(self, s):
        if len(s) <= 32:
            retval = super(URIMatcher, self).sub(s)
        else:
            retval = self.long_url_sub(s)
        return retval

    @staticmethod
    def add_zws(s):
        'Add zero-width spaces to the string'
        retval = ''
        for c in s:
            if c in punctuation:
                retval += ('&#8203;' + c)
            else:
                retval += c
        return retval

    def long_url_sub(self, s):
        m = self.re.match(s)
        gd = m.groupdict()
        brokenRest = self.add_zws(gd['rest'])
        c = '{leading}{protocol}<b>{host}</b>{rest}{trailing}'
        content = c.format(leading=gd['leading'], protocol=gd['protocol'], host=gd['host'],
                           rest=brokenRest, trailing=gd['trailing'])
        if len(s) > 64:
            r = '<a class="small" href="{0}">{1}</a>'
        else:
            r = '<a href="{0}">{1}</a>'
        url = '{0}{1}{2}'.format(gd['protocol'], gd['host'], gd['rest'])
        retval = r.format(url, content)
        return retval

#: Turn URIs (both ``http`` and ``https``) into clickable
#: links. If the link is particularly long (over 64 characters)
#: then small text will be used (``<a class="small"``). Leading and
#: trailing characters (like parenthesis) will be used in the
#: *link text* while just the URL will be used for the *link target*.
uriMatcher = URIMatcher()
